### PR TITLE
[CI] Reset global config in case of corrupt .gitconfig

### PR DESCRIPTION
**Pull request by amd-justchen**
## Motivation

Resolve recent issues with corrupted git configs as noted in https://github.com/ROCm/TheRock/issues/2929, https://github.com/ROCm/TheRock/issues/2956 to increase runner stability.

## Technical Details

Malformed global `.gitconfig` file causes `git` commands such as `git config --global ...` to fail and stop test workflows. Due to using persistent runners, workflows will repeatedly fail as a result, so add steps to clean up the git config in `setup_test_environment` action which is used to setup test workflows in all repos.

## Test Plan

Observe CI for test runs on all potentially affected Windows machines, but most notably:
- `windows-strix-halo-gpu-rocm-2`
- `windows-strix-halo-gpu-rocm-4`
- `windows-strix-halo-gpu-rocm-5`
- `windows-strix-halo-gpu-rocm-8`
## Test Result

Initial comments here about results https://github.com/ROCm/TheRock/pull/2967#issuecomment-3759686849

Followup comment about refactored script results being successful https://github.com/ROCm/TheRock/pull/2967#issuecomment-3762861165

## Submission Checklist

- [X] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Comment by amd-justchen** [comment]
It seems it is functioning as intended as `windows-strix-halo-gpu-rocm-4` is completing the `Adjust git config` step:
https://github.com/ROCm/TheRock/actions/runs/21055530432/job/60558265094#step:6:1
after removing `~/.gitconfig` (but it may also have been fixed prior to this test job)

However, when it is the very first run where `~/.gitconfig` is corrupt and needs to be cleaned up there is an edge case where steps using git commands prior to the `Run setup test environment workflow`  step would fail, such as on `windows-strix-halo-gpu-rocm-8` in the `Fetch 'build_tools' from repository` and `Checkout Repository` steps:
https://github.com/ROCm/TheRock/actions/runs/21055530432/job/60558265096#step:2:31
https://github.com/ROCm/TheRock/actions/runs/21055530432/job/60558265096#step:4:29

As suspected, these official Github provided actions do have the capability to call [git config --global](https://github.com/actions/checkout/blob/8e8c483db84b4bee98b60c0593521ed34d9990e8/src/git-command-manager.ts#L251), but they are robust enough to switch to downloading the repository with the Github REST API:

```haskell
Getting Git version info
  Working directory is 'C:\runner\_work\TheRock\TheRock'
  Error: fatal: bad config line 1 in file C:/WINDOWS/system32/config/systemprofile/.gitconfig
The repository will be downloaded using the GitHub REST API
To create a local Git repository instead, add Git 2.18 or higher to the PATH
Downloading the archive
Writing archive to disk
Extracting the archive
```

I would assume since the repository was simply extracted, it may [not contain](https://github.com/actions/checkout/issues/335#issue-683140675) a `.git` folder and as a result the `Adjust git config` step would now fail for a new reason when `git config` attempts to access the local repo config as it does here https://github.com/ROCm/TheRock/actions/runs/21055530432/job/60558265096#step:6:21

```haskell
Run git config --global --add safe.directory $PWD
  git config --global --add safe.directory $PWD
  git config fetch.parallel 10
  ...
fatal: not in a git directory
Error: Process completed with exit code 128.
```

Ideally this `.gitconfig` cleanup should run before any use of git commands in any step, but just having it at all would at least allow the runner to self heal and succeed on subsequent runs.

File: .github/actions/setup_test_environment/action.yml [hunk 1]
```diff
@@ -24,6 +24,12 @@ inputs:
 runs:
   using: "composite"
   steps:
+    # Reset git config in case of corrupted and bloated global .gitconfig on persistent test runners (#2929)
+    - name: "Cleaning up existing global .gitconfig"
+      shell: bash
+      run: |
+        rm -f ~/.gitconfig
```
Comment (ScottTodd) [inline .github/actions/setup_test_environment/action.yml]: This does not seem like a good fix. We should not be unconditionally deleting files outside of the runner's working directory.

* Can this go in a "system health check" step/script that we put on select runner machines instead of in our workflow files?
* Can we detect a "corrupted and bloated global .gitconfig" somehow? How are those appearing in the first place?

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: .github/actions/setup_test_environment/action.yml (hunk 1 above)
Comment (lamikr) [reply .github/actions/setup_test_environment/action.yml]: - We can detect it at the moment at least in a way that if we try to update it with new value we will get error.
- There is another "more complicated" PR which try to read the existing values and compares them to new ones and then try to update them only if changed. In that code I catch the error value and logs. It seems to still have failed but at least I got now better better error message:

https://github.com/ROCm/TheRock/actions/runs/21050697872/job/60555717857?pr=2959

```
Run python build_tools/test_ci_git_config_value_updater.py $PWD 10
Updating .gitconfig directory keyword. Old value: '', new value: 'C:/runner/_work/TheRock/TheRock'.
git config safe.directory value update failed with error code: 128
trying git config again but now with --replace-all parameter to maintain compatibility
git config safe.directory value update failed with error code: 128
Error: Process completed with exit code 128.

```

I will try to modify the logic in this python script to delete the .gitconfig if it gets the error and then try to setup config values again.

- For the corruption, one possibility that Justin found out is that the git itself has crashed on some command and caused the config file to get corrupted. See https://github.com/ROCm/TheRock/pull/2959#issuecomment-3757952462

Comment (amd-justchen) [reply .github/actions/setup_test_environment/action.yml]: Discussion with Scott:

> Hi Scott! Thanks for the review and I see where you're coming from in terms of unconditionally deleting a file external to the repo. In terms of a health checker script on the runner, do you mean this? https://docs.github.com/en/actions/how-tos/manage-runners/self-hosted-runners/run-scripts
> 
> I think in general many of the windows runner issues are coming from being persistent runners where jobs leave the system in a bad state. Why the .gitconfig is being corrupted, I'm still not sure but I think it's reasonable in the context of giving each job a clean system state to run in, to start with no .gitconfig as it sets it up during the job anyway. 
> 
> With that in mind would it be okay to delete it? Or at least okay to delete it in a run script? This would require manual changes to every one of the dozens windows runners that we have (since we have no orchestration yet) which is why I was thinking about putting it in the setup test environment action. Ideally we would have a container-like isolation for each job that runs on a machine.

He gave confirmation on either the use of GitHub run script for the runner or at least use a script in a workflow step instead of `rm` directly in the workflow.

Comment (amd-justchen) [reply .github/actions/setup_test_environment/action.yml]: Implemented in https://github.com/ROCm/TheRock/pull/2967/commits/915d107143c8594f73270a16509ab31e7ddeea08

**Review by lamikr — APPROVED** [review]

**Comment by amd-justchen** [comment]
The `gfx1151`/`Strix Halo` runner queue has been quite long, but it finally completed the first job `Configure test matrix` on the GPU runner  `windows-strix-halo-gpu-rocm-2` which has been plagued by the corrupt `.gitconfig` issue, but notably:
1. During the job's [2nd step](https://github.com/ROCm/TheRock/actions/runs/21085333351/job/60653119457?pr=2967#step:2:23), `Fetch 'build_tools' from repository`, which does a sparse checkout (using `actions/checkout`)  for the `cleanup_processes.ps1` script, it had the usual error with git: `Error: fatal: bad config line 1 in file C:/windows/system32/config/systemprofile/.gitconfig` but it continued with the GitHub REST API fallback:
```haskell
Run actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8
Syncing repository: ROCm/TheRock
Getting Git version info
  Working directory is 'B:\actions-runner\_work\TheRock\TheRock\prejob'
  "C:\Program Files\Git\cmd\git.exe" version
  Error: fatal: bad config line 1 in file C:/windows/system32/config/systemprofile/.gitconfig
```
2. This still allows `cleanup_processes.ps1`, which now has the environment cleanup change, to run in the next step, `Pre-job cleanup processes on Windows`:
```js
[*] ==== Starting cleanup_processes.ps1 ====
[*] Checking if elevated: True | current user: NT AUTHORITY\SYSTEM
[*] Running as a Windows Service (NT AUTHORITY\*) - Cleaning up system environment...
[*] > Checking for "~/.gitconfig" file at: C:\windows\system32\config\systemprofile\.gitconfig
[*] >> .gitconfig file found, removing after logging contents:
fatal: bad config line 1 in file C:/windows/system32/config/systemprofile/.gitconfig
[+] >> Successfully removed .gitconfig
...
```
3. In the next step, `Checking out repository`, `actions/checkout` is used again but this time for the job's main checkout step, and more importantly is successfully able to run `git` commands as any faulty .gitconfig is removed.
```haskell
Run actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8
Syncing repository: ROCm/TheRock
Getting Git version info
  Working directory is 'B:\actions-runner\_work\TheRock\TheRock'
  "C:\Program Files\Git\cmd\git.exe" version
  git version 2.49.0.windows.1               <<<<<<<<<< GIT OUTPUT SUCCESS
Temporarily overriding HOME='B:\actions-runner\_work\_temp\888701e2-6347-4551-ab77-fc402bf1701f' before making global git config changes
Adding repository directory to the temporary git global config as a safe directory
"C:\Program Files\Git\cmd\git.exe" config --global --add safe.directory B:\actions-runner\_work\TheRock\TheRock
Deleting the contents of 'B:\actions-runner\_work\TheRock\TheRock'
Initializing the repository
Disabling automatic garbage collection
Setting up auth
Fetching the repository
Determining the checkout info
"C:\Program Files\Git\cmd\git.exe" sparse-checkout disable
"C:\Program Files\Git\cmd\git.exe" config --local --unset-all extensions.worktreeConfig
Checking out the ref
"C:\Program Files\Git\cmd\git.exe" log -1 --format=%H
5619b97840adc0784c7b4475b022485563a39f71
```

Given the successful `git` invocations, I believe the cleanup is working as intended to workaround reoccurring .gitconfig corruption and will merge without needing the remaining jobs to finish, such as `Sanity ROCM Test` (which is waiting in the long queue for Strix Halo runners).

---
#### Note:
As part of `actions/checkout` there is also a `Post Checkout Repository` which runs at the end of the job but, right before it `cleanup_processes.ps1` is re-run in the `Post Run setup test environment workflow`, which will clear the .gitconfig again, but I noticed it is copied out. From what I understand there shouldn't be an issue and it's merely appending to it.

```haskell
Post job cleanup.
"C:\Program Files\Git\cmd\git.exe" version
git version 2.50.1.windows.1
Copying 'C:\windows\system32\config\systemprofile\.gitconfig' to 'C:\runner\_work\_temp\484e7ae4-88d6-478f-955c-5861fcbf898b\.gitconfig'
Temporarily overriding HOME='C:\runner\_work\_temp\484e7ae4-88d6-478f-955c-5861fcbf898b' before making global git config changes
```
